radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -69,11 +69,28 @@ public VertexStore(
 		this.lastCommittedVertex.onNext(genesisVertex);
 	}
 
-	public void syncToQC(QuorumCertificate qc) throws SyncException {
+	public void syncToQC(QuorumCertificate qc, VertexSupplier vertexSupplier) throws SyncException {
 		final Vertex vertex = vertices.get(qc.getProposed().getId());
 		if (vertex == null) {
-			// TODO: actual syncing
-			throw new SyncException(qc);
+			if (!vertices.containsKey(qc.getParent().getId())) {
+				// Too far behind
+				// TODO: more syncing
+				throw new SyncException(qc);
+			}
+
+			final Vertex proposedVertex;
+			try {
+				// TODO: remove blocking
+				proposedVertex = vertexSupplier.getVertex(qc.getProposed().getId()).blockingGet();
+			} catch (Exception e) {
+				throw new SyncException(qc, e);
+			}
+
+			try {
+				this.insertVertex(proposedVertex);
+			} catch (VertexInsertionException e) {
+				throw new IllegalStateException("Should not go here.");
```

Comment:
Should not go here because of the earlier check (line 75)? Maybe add a comment to that effect and include the exception as a cause to the `IllegalStateException`, the trigger for `VertexInsertionException` may change in the future

Reply:
This has been resolved, yet the code wasn't changed here - can you clarify @talekhinezh?

Reply:
missed "cause" comment